### PR TITLE
Add PR Agent review, describe, and improve workflow on CI

### PR DESCRIPTION
**Pull request by PeterDaveHello**
### **User description**
Generate PR summaries, review feedback, and GitHub suggestion blocks for same-repository PR updates and maintainer-requested reviews.

Allow trusted maintainers to trigger PR Agent on fork PRs with slash commands without switching the workflow to pull_request_target.

Set PR Agent to use gpt-5.4-mini with xhigh reasoning to balance review quality and API cost.

___

### **PR Type**
Enhancement

___

### **Description**
- Runs parallel `describe`/`review`/`improve` jobs on same-repo PRs

- Accepts trusted slash commands on open PRs

- Cancels stale reruns with concurrency groups

- Pins `pr-agent` and configures model settings

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["pull_request events"] -- "same-repo, non-draft" --> B["Parallel describe, review, improve jobs"]
  C["Trusted issue_comment commands"] -- "open PRs" --> D["Manual slash-command run"]
  B -- "uses" --> E["Pinned pr-agent image + model config"]
  D -- "uses" --> E
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pr-agent.yml</strong><dd><code>Add PR Agent workflow automation and commands</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/workflows/pr-agent.yml

<ul><li>Adds <code>pull_request</code> and <code>issue_comment</code> triggers.<br> <li> Runs <code>describe</code>/<code>review</code>/<code>improve</code> jobs in parallel for same-repo, non-draft <br>PRs.<br> <li> Accepts trusted slash commands on open PRs, including forks, and <br>cancels stale reruns.<br> <li> Pins <code>pr-agent</code> and configures model, timeout, and suggestion settings.</ul>

</details>

  </td>
  <td><a href="https://github.com/thlorenz/doctoc/pull/367/files#diff-69fd3473f15a98f47918e81bc4d0ca86a71131f21db13b39bebf76d27cf483b4">+93/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: .github/workflows/pr-agent.yml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      pull-requests: write
+
+    steps:
+      - name: Run PR Agent review
+        uses: docker://pragent/pr-agent@sha256:4ec7bac814050a1bc8c96ab2fab6b7b0f65df0049a5ec43f3fee1a0b551c28ca
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          OPENAI_KEY: ${{ secrets.OPENAI_KEY }}
+          config.model: "gpt-5.4-mini"
+          config.reasoning_effort: "xhigh"
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply .github/workflows/pr-agent.yml]: I do not think this change is needed. The latest PR Agent GitHub Action docs use the same dotted env key format (`config.model`, `github_action_config.auto_review`, and `pr_code_suggestions.*`) and explicitly say env vars may use dots (`.`) or double underscores (`__`) to separate sections and keys: https://github.com/The-PR-Agent/pr-agent/blob/main/docs/docs/installation/github.md#environment-variable-format

The Actions run also confirms these keys were applied at runtime: PR Agent resolved `"model": "gpt-5.4-mini"` and `"reasoning_effort": "xhigh"`, then logged `Generating prediction with gpt-5.4-mini` and `Using reasoning_effort='xhigh' for GPT-5 model`.

So the dotted lowercase env vars are not being ignored.

File: .github/workflows/pr-agent.yml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+          github_action_config.auto_review: "true"
+          github_action_config.auto_describe: "true"
+          github_action_config.auto_improve: "true"
+          pr_code_suggestions.commitable_code_suggestions: "true"
+          github_action_config.pr_actions: '["opened", "reopened", "ready_for_review", "review_requested", "synchronize"]'
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply .github/workflows/pr-agent.yml]: I do not think this change is needed. The latest PR Agent GitHub Action docs use the same dotted env key format (`config.model`, `github_action_config.auto_review`, and `pr_code_suggestions.*`) and explicitly say env vars may use dots (`.`) or double underscores (`__`) to separate sections and keys: https://github.com/The-PR-Agent/pr-agent/blob/main/docs/docs/installation/github.md#environment-variable-format

The Actions run also confirms these keys were applied at runtime: PR Agent resolved `"model": "gpt-5.4-mini"` and `"reasoning_effort": "xhigh"`, then logged `Generating prediction with gpt-5.4-mini` and `Using reasoning_effort='xhigh' for GPT-5 model`.

So the dotted lowercase env vars are not being ignored.

File: .github/workflows/pr-agent.yml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        github.event_name == 'pull_request' &&
+        github.event.sender.type != 'Bot' &&
+        github.event.pull_request.draft == false &&
+        github.event.pull_request.head.repo.full_name == github.repository
+      ) ||
+      (
+        github.event_name == 'issue_comment' &&
+        github.event.sender.type != 'Bot' &&
+        github.event.issue.pull_request != null &&
+        contains(fromJSON('["OWNER", "MEMBER", "COLLABORATOR"]'), github.event.comment.author_association) &&
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply .github/workflows/pr-agent.yml]: Not a good idea to add FIRST_TIMER / FIRST_TIME_CONTRIBUTOR, but CONTRIBUTOR could be acceptable.

File: .github/workflows/pr-agent.yml [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  cancel-in-progress: true
+
+permissions:
+  contents: read
+  issues: write
+  pull-requests: write
+
+env:
+  GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+  OPENAI_KEY: ${{ secrets.OPENAI_KEY }}
```
Comment (AndrewSouthpaw) [inline .github/workflows/pr-agent.yml:18]: where is the OpenAI key coming from? Whose account? This seems like a blocker.

File: .github/workflows/pr-agent.yml [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+            auto_improve: "false"
+          - tool: review
+            auto_describe: "false"
+            auto_review: "true"
+            auto_improve: "false"
+          - tool: improve
+            auto_describe: "false"
+            auto_review: "false"
+            auto_improve: "true"
+
```
Comment (AndrewSouthpaw) [inline .github/workflows/pr-agent.yml:55]: this matrix seems like unnecessary overhead, just run them all together in one?

File: .github/workflows/pr-agent.yml [hunk 6]
```diff
@@ -0,0 +1,92 @@
+name: PR Agent
```
Comment (AndrewSouthpaw) [inline .github/workflows/pr-agent.yml:1]: IDK what those trailing comments are about (e.g. [here](https://github.com/thlorenz/doctoc/pull/367#issuecomment-4793800062)) from the bot but they're annoying and should be prevented.

File: .github/workflows/pr-agent.yml [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        contains(fromJSON('["/review", "/describe", "/improve", "/ask"]'), github.event.comment.body) ||
+        startsWith(github.event.comment.body, '/review ') ||
+        startsWith(github.event.comment.body, fromJSON('"/review\n"')) ||
+        startsWith(github.event.comment.body, '/describe ') ||
+        startsWith(github.event.comment.body, fromJSON('"/describe\n"')) ||
+        startsWith(github.event.comment.body, '/improve ') ||
+        startsWith(github.event.comment.body, fromJSON('"/improve\n"')) ||
+        startsWith(github.event.comment.body, '/ask ') ||
+        startsWith(github.event.comment.body, fromJSON('"/ask\n"'))
+      )
```
Comment (AndrewSouthpaw) [inline .github/workflows/pr-agent.yml:81]: seems overkill, let's just support one command like `/review`

**Review by AndrewSouthpaw — CHANGES_REQUESTED** [review]

File: .github/workflows/pr-agent.yml [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+            auto_review: "true"
+            auto_improve: "false"
+          - tool: improve
+            auto_describe: "false"
+            auto_review: "false"
+            auto_improve: "true"
+
+    steps:
+      - name: Run PR Agent ${{ matrix.tool }}
+        uses: &pr-agent-image docker://pragent/pr-agent@sha256:4ec7bac814050a1bc8c96ab2fab6b7b0f65df0049a5ec43f3fee1a0b551c28ca
```
Comment (AndrewSouthpaw) [inline .github/workflows/pr-agent.yml]: use a version tag, the sha pin is hard to interpret.

File: .github/workflows/pr-agent.yml (hunk 4 above)
Comment (PeterDaveHello) [reply .github/workflows/pr-agent.yml:18]: Currently it's mine, that's the reason why I want to keep it using a more affordable model.

File: .github/workflows/pr-agent.yml (hunk 7 above)
Comment (PeterDaveHello) [reply .github/workflows/pr-agent.yml:81]: "/improve" should be its core function, otherwise it won't provide code suggestions.

File: .github/workflows/pr-agent.yml (hunk 5 above)
Comment (PeterDaveHello) [reply .github/workflows/pr-agent.yml:55]: The goal of doing it this way is to trigger different tasks in parallel, effectively speeding up the entire workflow. This ensures that both contributors and reviewers don't have to spend too much time waiting for automated feedback.

While the implementation will certainly be slightly more complex, I believe it's worth considering given the value of the time saved waiting.

File: .github/workflows/pr-agent.yml [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+          github_action_config.auto_review: ${{ matrix.auto_review }}
+          github_action_config.auto_describe: ${{ matrix.auto_describe }}
+          github_action_config.auto_improve: ${{ matrix.auto_improve }}
+
+  manual:
+    if: >-
+      github.event_name == 'issue_comment' &&
+      github.event.sender.type != 'Bot' &&
+      github.event.issue.pull_request != null &&
+      contains(fromJSON('["OWNER", "MEMBER", "COLLABORATOR", "CONTRIBUTOR"]'), github.event.comment.author_association) &&
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply .github/workflows/pr-agent.yml:70]: Added the open-PR guard to block commands on closed PR threads.

Keeping CONTRIBUTOR is intentional: it allows prior contributors to trigger PR Agent on fork PRs, without allowing first-time or arbitrary commenters. No PR code is checked out or executed.

File: .github/workflows/pr-agent.yml (hunk 8 above)
Comment (PeterDaveHello) [reply .github/workflows/pr-agent.yml]: yeah, I missed that :sweat_smile:

File: .github/workflows/pr-agent.yml [hunk 10]
```diff
@@ -0,0 +1,92 @@
+name: PR Agent
+
+on:
+  pull_request:
+    branches: ["master"]
+    types: [opened, reopened, ready_for_review, review_requested, synchronize]
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply .github/workflows/pr-agent.yml:6]: Should be the general cases.

File: .github/workflows/pr-agent.yml [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+  issue_comment:
+    types: [created]
+
+concurrency:
+  # PR events group by PR so newer pushes replace stale auto runs.
+  # issue_comment events use run_id here; manual commands are grouped below.
+  group: ${{ github.workflow }}-${{ github.event_name }}-${{ github.event_name == 'pull_request' && github.event.pull_request.number || github.run_id }}
+  cancel-in-progress: true
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply .github/workflows/pr-agent.yml:14]: It's fine for "manual" job.

File: .github/workflows/pr-agent.yml (hunk 8 above)
Comment (PeterDaveHello) [reply .github/workflows/pr-agent.yml]: Fixed.

File: .github/workflows/pr-agent.yml (hunk 6 above)
Comment (PeterDaveHello) [reply .github/workflows/pr-agent.yml:1]: Fixed.

File: .github/workflows/pr-agent.yml (hunk 5 above)
Comment (AndrewSouthpaw) [reply .github/workflows/pr-agent.yml:55]: I hear that, and also my experience is that I can give AI a prompt like "look over the PR description, code, and ways to improve the code" and it'll give an answer equivalently fast. I don't think we're saving a worthwhile amount of time. Do you have numbers on the amount of time saved?

File: .github/workflows/pr-agent.yml (hunk 7 above)
Comment (AndrewSouthpaw) [reply .github/workflows/pr-agent.yml:81]: That suggestion is assuming we collapse into a single function; otherwise supporting all the slash commands makes sense.

**Review by AndrewSouthpaw — CHANGES_REQUESTED** [review]
I'm still a "no" for now. AI slop is running rampant, and it's especially bad with lower thinking models. AI review bots are particularly infuriating, because they slowly dole out one quasi-useful bit of feedback at a time, forcing you into a demoralizing endless loop of iterations. 

I have to deal with enough slop at my workplace, I don't need to see it here as well. I am not fussed to tell my higher-quality model to review incoming PRs, then take the feedback, QA it, and then publish.

**Comment by PeterDaveHello** [comment]
Appreciate the continued review, @AndrewSouthpaw.

To clarify cost and ownership first: any OpenAI API usage for this workflow would be covered by me personally using my own official OpenAI API credits, not by the project. If this moves forward, that responsibility can be documented explicitly.

I understand the concern about the quality of AI-generated review output. That is a valid risk, but one that can be managed through scope control, trigger restrictions, model selection, and final maintainer review. At the same time, AI-assisted code review has become a fairly common part of modern development workflows, including GitHub’s own Copilot review capabilities. The models and tools have continued to improve significantly over the past few years, and they can now provide faster first-pass feedback as well as an additional review perspective that may catch issues a single human reviewer might overlook.

AI review is not perfect or authoritative. Like any reviewer, it can be wrong, incomplete, or overly confident, so maintainer judgment remains essential. However, it can serve as a consistent and less fatigue-prone first pass, particularly for repetitive checks, edge cases, documentation consistency, test behavior, and workflow details. This repository has already used GitHub Copilot for code review on multiple PRs in the past, with personal credits, and it provided useful feedback in those areas.

The intent of this PR is to make that pattern more controlled: a maintainer-triggered assistant for PR describe, review, and improve workflows, with clear guardrails around permissions, cost, and ownership. It is not meant to replace human review, require anyone to rely on the output, or allow untrusted actors to trigger paid runs.

Based on the review so far, it seems the remaining concern is with the overall direction rather than the specific workflow details. If there are concrete guardrails that would address the concerns raised here, I’m open to narrowing the PR accordingly. If the objection is to having AI-assisted PR review in the repository at all, it may be better to pause further implementation changes on this PR and first seek broader maintainer alignment on that direction.
